data-prepper-plugins/kafka-connect-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafkaconnect/source/mongoDB/MongoDBPartitionCreationSupplier.java

```diff
@@ -0,0 +1,126 @@
+                                                        final Map<String, Object> partitionedCollections) {
+        return mongoDBConfig.getCollections()
+                .stream()
+                .map(MongoDBConfig.CollectionConfig::getCollectionName)
+                .filter(collectionName -> !partitionedCollections.containsKey(collectionName))
+                .collect(Collectors.toList());
+    }
+
+    private List<PartitionIdentifier> buildPartitions(final String collectionName) {
+        List<PartitionIdentifier> collectionPartitions = new ArrayList<>();
+        List<String> collection = List.of(collectionName.split("\\."));
+        if (collection.size() < 2) {
+            throw new IllegalArgumentException("Invalid Collection Name. Must as db.collection format");
+        }
+        try (MongoClient mongoClient = MongoDBHelper.getMongoClient(mongoDBConfig)) {
+            MongoDatabase db = mongoClient.getDatabase(collection.get(0));
+            MongoCollection<Document> col = db.getCollection(collection.get(1));
+            int chunkSize = this.mongoDBConfig.getExportConfig().getItemsPerPartition();
+            FindIterable<Document> startIterable = col.find()
+                    .projection(new Document("_id", 1))
+                    .sort(new Document("_id", 1))
+                    .limit(1);
+            while (true) {
+                try (MongoCursor<Document> startCursor = startIterable.iterator()) {
+                    if (!startCursor.hasNext()) {
+                        break;
+                    }
+                    Document startDoc = startCursor.next();
+                    Object gteValue = startDoc.get("_id");
+                    String className = gteValue.getClass().getName();
+
+                    // Get end doc
+                    Document endDoc = startIterable.skip(chunkSize - 1).limit(1).first();
+                    if (endDoc == null) {
+                        // this means we have reached the end of the doc
+                        endDoc = col.find()
+                                .projection(new Document("_id", 1))
+                                .sort(new Document("_id", -1))
+                                .limit(1)
+                                .first();
+                    }
+                    if (endDoc == null) {
+                        break;
+                    }
+
+                    Object lteValue = endDoc.get("_id");
+                    LOG.info("Chunk of " + collectionName + ": {gte: " + gteValue.toString() + ", lte: " + lteValue.toString() + "}");
+                    collectionPartitions.add(
+                            PartitionIdentifier
+                                    .builder()
+                                    .withPartitionKey(String.format(DOCUMENTDB_PARTITION_KEY_FORMAT, collectionName, gteValue, lteValue, className))
+                                    .build());
+
+                    startIterable = col.find(Filters.gt("_id", lteValue))
```

Comment:
can this be null ? What is the value when end of doc/collection is reached ?

Reply:
Yes, it can be null. That is why there is a break in code above. If it reaches the end, it doesn't have next, then break.
```
        try (MongoCursor<Document> startCursor = startIterable.iterator()) {
                    if (!startCursor.hasNext()) {
                        break;
                    }
                   ....
     }
```